src/comment_worker.py

```diff
@@ -568,28 +599,82 @@ def promote(self, sess, comment, investor, to_promote):
             first()
 
         if user.firm_role == "":
+            if (investor.firm_role == "") or (investor.firm_role == "assoc"):
+                return comment.reply_wrap(message.not_ceo_or_exec_org)
+
+            max_assocs = max_assocs_for_rank(firm.rank)
+            if firm.assocs >= max_assocs:
+                return comment.reply_wrap(message.modify_promote_assocs_full(firm))
+
+            user.firm_role = "assoc"
+            firm.assocs += 1
+
+        elif user.firm_role == "assoc":
+            if (investor.firm_role != "ceo") or (investor.firm_role != "coo"):
+                return comment.reply_wrap(message.not_ceo_or_coo_org)
+
             max_execs = max_execs_for_rank(firm.rank)
             if firm.execs >= max_execs:
-                return comment.reply_wrap(message.modify_promote_full(firm))
+                return comment.reply_wrap(message.modify_promote_execs_full(firm))
 
             user.firm_role = "exec"
+            firm.assocs -= 1
             firm.execs += 1
+
         elif user.firm_role == "exec":
+            if investor.firm_role != "ceo":
+                return comment.reply_wrap(message.not_ceo_org)
+
+            if firm.cfo >= 1
+                return comment.reply_wrap(message.promote_cfo_full_org)
+
+            user.firm_role = "cfo"
+            firm.execs -= 1
+            firm.cfo += 1
+
+        elif user.firm_role == "cfo"
+            if investor.firm_role != "ceo":
+                return comment.reply_wrap(message.not_ceo_org)
+
+            if firm.coo >= 1
+                return comment.reply_wrap(message.promote_coo_full_org)
+
+            user.firm_role = "coo"
+            firm.cfo -= 1
+            firm.coo += 1
+
+        elif user.firm_role == "coo"
+            if investor.firm_role != "ceo":
+                return comment.reply_wrap(message.not_ceo_org)
+
             # Swapping roles
-            investor.firm_role = "exec"
+            investor.firm_role = "coo"
             user.firm_role = "ceo"
 
         # Updating the flair in subreddits
         flair_role_user = ''
         if user.firm_role == "ceo":
             flair_role_user = "CEO"
-        else:
+        elif user.firm_role == "coo":
+            flair_role_user = "COO"
+        elif user.firm_role == "cfo":
+            flair_role_user = "CFO"
+        elif user.firm_role == "exec":
             flair_role_user = "Executive"
+        elif user.firm_role == "assoc":
+            flair_role_user = "Associate"
+
         flair_role_investor = ''
         if investor.firm_role == "ceo":
             flair_role_investor = "CEO"
-        else:
+        elif investor.firm_role == "coo":
+            flair_role_investor = "COO"
+        elif investor.firm_role == "cfo":
+            flair_role_investor = "CFO"
+        elif investor.firm_role == "exec":
             flair_role_investor = "Executive"
+        elif investor.firm_role == "assoc":
+            flair_role_investor = "Associate"
```

Comment:
Lines 590 through 677: Add new roles to "!promote" command as well as adjusting role permissions